stacks-common/src/util/lru_cache.rs

```diff
@@ -0,0 +1,596 @@
+            .insert_with_dirty(1, 1, false)
+            .expect("cache corrupted")
+            .is_none());
+        assert!(cache
+            .insert_with_dirty(2, 2, true)
+            .expect("cache corrupted")
+            .is_none());
+        assert!(cache
+            .insert_with_dirty(3, 3, true)
+            .expect("cache corrupted")
+            .is_none());
+
+        let mut flushed = Vec::new();
+        cache
+            .flush(|k, v| {
+                flushed.push((*k, v));
+                Ok::<(), ()>(())
+            })
+            .expect("cache corrupted")
+            .expect("flush failed");
+
+        assert_eq!(flushed, [(3, 3), (2, 2)]);
+    }
+
+    /// Simple LRU implementation for testing
+    pub struct SimpleLRU {
+        pub cache: Vec<Node<u32, u32>>,
+        capacity: usize,
+    }
+
+    impl SimpleLRU {
+        pub fn new(capacity: usize) -> Self {
+            SimpleLRU {
+                cache: Vec::with_capacity(capacity),
+                capacity,
+            }
+        }
+
+        pub fn insert(&mut self, key: u32, value: u32, dirty: bool) {
+            if let Some(pos) = self.cache.iter().position(|x| x.key == key) {
+                self.cache.remove(pos);
+            } else if self.cache.len() == self.capacity {
+                self.cache.remove(0);
+            }
+            self.cache.push(Node {
+                key,
+                value,
+                dirty,
+                next: 0,
+                prev: 0,
+            });
+        }
+
+        pub fn get(&mut self, key: u32) -> Option<u32> {
+            if let Some(pos) = self.cache.iter().position(|x| x.key == key) {
+                let node = self.cache.remove(pos);
+                let value = node.value;
+                self.cache.push(node);
+                Some(value)
+            } else {
+                None
+            }
+        }
+
+        pub fn flush<E>(&mut self, mut f: impl FnMut(&u32, u32) -> Result<(), E>) -> Result<(), E> {
+            for node in self.cache.iter_mut().rev() {
+                if node.dirty {
+                    f(&node.key, node.value)?;
+                }
+                node.dirty = false;
+            }
+            Ok(())
+        }
+    }
+}
+
+#[cfg(test)]
+mod property_tests {
+    use proptest::prelude::*;
+
+    use super::tests::SimpleLRU;
+    use super::*;
+
+    #[derive(Debug, Clone)]
+    enum CacheOp {
+        Insert(u32, u32),
+        Get(u32),
+        InsertClean(u32, u32),
+        Flush,
+    }
+
+    prop_compose! {
+        fn arbitrary_op()(op_type in 0..4, key in 0..100u32, value in 0..1000u32) -> CacheOp {
+            match op_type {
+                0 => CacheOp::Insert(key, value),
+                1 => CacheOp::Get(key),
+                2 => CacheOp::InsertClean(key, value),
+                _ => CacheOp::Flush,
+            }
+        }
+    }
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(1_000_000))]
```

Comment:
Since we went with the safer fallible option, I'm thinking we can let these tests just run the default number of cases (256) and be much faster. Any complaints?

Reply:
Changed in d64b3679a149e31a2fb71300a954c858d9ccc741.